core/system/src/context.c

```diff
@@ -0,0 +1,266 @@
+        HALT_ERROR(SANITY_CHECK_FAILED, "Context state stack overflow");
+    }
+
+    /* Push the source box state to the state stack. */
+    g_context_previous_states[g_context_p].type = context_type;
+    g_context_previous_states[g_context_p].src_id = src_id;
+    g_context_previous_states[g_context_p].src_sp = src_sp;
+    ++g_context_p;
+
+    /* Update the current state of the source box. */
+    g_context_current_states[src_id].sp = src_sp;
+}
+
+/** Pops the previous state from the state stack.
+ *
+ * @internal
+ *
+ * @warning This function trusts all the arguments that are passed to it. Input
+ * verification should be performed by the caller.
+ *
+ * @returns the pointer to the previous box context state. */
+static TContextPreviousState * context_state_pop(void)
+{
+    /* Check that the state stack does not underflow. */
+    if (!g_context_p) {
+        HALT_ERROR(SANITY_CHECK_FAILED, "Context state stack underflow");
+    }
+
+    /* Pop the source box state and return it to the caller. */
+    --g_context_p;
+    return &g_context_previous_states[g_context_p];
+}
+
+/* Returns a pointer to the previous box context state. */
+TContextPreviousState * context_state_previous(void)
+{
+    if (!g_context_p) {
+        /* There is no previous state */
+        return NULL;
+    } else {
+        return &g_context_previous_states[g_context_p - 1];
+    }
+}
+
+/* Forges a new exception stack frame and copy arguments from an old one. */
+uint32_t context_forge_exc_sf(uint32_t src_sp, uint8_t dst_id, uint32_t dst_fn, uint32_t dst_lr, uint32_t xpsr, int nargs)
+{
+    uint32_t dst_sp;
+    uint32_t exc_sf_alignment;
+
+    /* Destination box: Gather information from the current state. */
+    /* Note: Here we allow source and destination box IDs to be the same, as
+     * this function is used also for IRQs. */
+    if (dst_id != dst_id) {
+        dst_sp = g_context_current_states[dst_id].sp;
+    } else {
+        dst_sp = src_sp;
+    }
+
+    /* Forge an exception stack frame in the destination box stack. */
+    exc_sf_alignment = (dst_sp & 0x4) ? 1 : 0;
+    dst_sp -= (CONTEXT_SWITCH_EXC_SF_SIZE + exc_sf_alignment);
+
+    /* Populate the new exception stack frame. */
+
+    /* r0 - r3. Unused slots are explicitly not set. The values that will be
+     * unstacked belong to the destination stack so are not leaked from another
+     * context. */
+    memcpy((void *) dst_sp, (void *) src_sp, sizeof(uint32_t) * nargs);
+
+    /* r12 */
+    ((uint32_t *) dst_sp)[4] = 0;
+
+    /* lr */
+    ((uint32_t *) dst_sp)[5] = dst_lr;
+
+    /* Return value. This is always OR-ed with 1 (Thumb bit). */
+    ((uint32_t *) dst_sp)[6] = dst_fn | 1;
+
+    /* Stacked xPSR register. This always includes information on the stack
+     * frame alignment, regardless of the input from the caller. */
+    ((uint32_t *) dst_sp)[7] = xpsr | (exc_sf_alignment << 9);
+
+    /* Return pointer to the newly created stack frame. */
```

Comment:
Changed the comment to make clearer. @meriac and I discussed this some time ago and concluded that for security purposes is better to have pointer stored and treated as `uint32_t`, which prevents us from using the as a pointers (de-reference it) by mistake.

In general, we are not supposed to cast-and-deference these pointers directly, but instead should use the unprivileged-access functions `vmpu_unpriv_{read,write}` which result in `*t` instructions. In this case this is not possible as we are accessing elements of a box different from the currently active one.